lib/hash_ring/ets.ex

```diff
@@ -252,21 +288,47 @@ defmodule ExHashRing.HashRing.ETS do
       {:reply, {:error, :node_exists}, state}
     else
       nodes = [{node_name, num_replicas} | nodes]
-      {:reply, {:ok, nodes}, rebuild(%{state | nodes: nodes})}
+      {:reply, {:ok, nodes}, update_nodes(state, nodes)}
+    end
+  end
+
+  def handle_call({:add_nodes, nodes}, _from, %__MODULE__{} = state) do
+    nodes = transform_nodes(nodes, state.default_num_replicas)
+
+    existing_nodes = Enum.filter(nodes, fn {name, _} ->
+      has_node_with_name?(state.nodes, name)
+    end)
+
+    if Enum.empty?(existing_nodes) do
+      nodes = nodes ++ state.nodes
+      {:reply, {:ok, nodes}, update_nodes(state, nodes)}
+    else
+      {:reply, {:error, :node_exists}, state}
     end
   end
 
   def handle_call({:remove_node, node_name}, _from, %{nodes: nodes} = state) do
     if has_node_with_name?(nodes, node_name) do
       nodes = Enum.reject(nodes, fn {existing_node, _} -> existing_node == node_name end)
-      {:reply, {:ok, nodes}, rebuild(%{state | nodes: nodes})}
+      {:reply, {:ok, nodes}, update_nodes(state, nodes)}
+    else
+      {:reply, {:error, :node_not_exists}, state}
+    end
+  end
+
+  def handle_call({:remove_nodes, node_names}, _from, %__MODULE__{} = state) do
+    unknown_nodes = Enum.reject(node_names, &has_node_with_name?(state.nodes, &1))
+
+    if Enum.empty?(unknown_nodes) do
+      nodes = Enum.reject(state.nodes, fn {name, _} -> name in node_names end)
```

Comment:
Enum.count/1 is O(n) as it has to traverse the Enum, `count` is always O(n) `size` is always O(1).

I'm not sure if I follow the logic of the suggested alternative.  I can find the filtered nodes, I guess if the length of the filtered nodes is equal to the length of the state.nodes - length of the node_names then I would know that there are no unknowns.  

Since I actually don't care what nodes are unknown, just that there are unknown nodes, I could use Enum.any?/2 instead.

Reply:
Switched the check to use Enum.any?/2 in https://github.com/discord/ex_hash_ring/pull/10/commits/173fbda6fff6a11075af73ad466eed548991667b